File: dmoj/ccc/c++/2024/J5_Harvest_Waterloo.cpp
import sys
sys.setrecursionlimit(100003)
R, C = int(input()), int(input())
g = [list(input()) for i in range(R)]
r, c = int(input()), int(input())
ans = 0
d = {'S': 1, 'M': 5, 'L': 10}
def fun(r, c):
    global ans
    ans += d[g[r][c]]
    g[r][c] = '*'
    if r - 1 >= 0 and g[r - 1][c] != '*': fun(r - 1, c)
    if r + 1 < R and g[r + 1][c] != '*': fun(r + 1, c)
    if c - 1 >= 0 and g[r][c - 1] != '*': fun(r, c - 1)
    if c + 1 < C and g[r][c + 1] != '*': fun(r, c + 1)

fun(r, c)
print(ans)

File: CCC/python/2024/jr/J5 Harvest Waterloo.py
# recursive solution with a DFS
import sys 
input=sys.stdin.readline
sys.setrecursionlimit(10000)
rows = int(input())
cols = int(input())
grid = [list(input()) for i in range(rows)]
y=int(input())
x=int(input())
values={'S':1, 'M': 5, 'L': 10}
total=0

def go(ROW,COL):
    global total
    total += values[grid[ROW][COL]]
    grid[ROW][COL]='*'

    if COL + 1 < cols and grid[ROW][COL+1] != '*':
        go(ROW,COL+1)
    if COL - 1 >=0 and grid[ROW][COL-1] != '*':
        go(ROW,COL-1)
    if ROW+1 < rows and grid[ROW+1][COL] != '*':
        go(ROW+1,COL)
    if ROW-1 >= 0 and grid[ROW-1][COL] != '*':
        go(ROW-1,COL)

go(y,x)
print(total)

# iteratve solution with a BFS
r=int(input())
c=int(input())
patch=[input() for _ in range(r)]
pumpkin_values = {'S': 1, 'M': 5, 'L': 10}
total=0
directions=[(0,1),(1,0),(0,-1),(-1,0)]
a=int(input())
b=int(input())
q=[(a,b)]
visited=set([(a,b)])
while q:
    x,y=q.pop(0)
    if patch[x][y] in pumpkin_values:
        total += pumpkin_values[patch[x][y]]
    
    for dx,dy in directions:
        nx,ny = dx+x,dy+y
        if 0 <= nx < r and 0 <= ny < c and patch[nx][ny] != '*' and (nx,ny) not in visited:
            q.append((nx,ny))
            visited.add((nx,ny))
print(total)
